Privacy Enhancement for Polkadot Extension

nodi-andy opened the pull request:
### Project Abstract

> Please provide a brief description of your project here summarising key points (1-2 paragraphs).
>
> If your application is a follow-up to a previous grant, please mention which one in the first line of the abstract and include a link to previous pull requests if applicable.

### For which grant [level](https://github.com/w3f/Grants-Program#level_slider-levels) are you applying? 
- [x] **Level 1**:  Up to $10,000, 2 approvals
- [ ] **Level 2**:  Up to $50,000, 3 approvals
- [ ] **Level 3**:  Unlimited, 5 approvals (for > $100k Web3 Foundation Council approval)

### Application Checklist

- [x] The [application template](https://github.com/w3f/Grants-Program/blob/master/applications/application-template.md) has been copied, renamed ( `project_name.md`) and updated.
- [ ] A BTC or Ethereum (DAI/USDT) address for the payment of the milestones is provided inside the application.
- [ ] I have read and acknowledged the [terms and conditions](https://github.com/w3f/Grants-Program/blob/master/docs/T&Cs.md).
- [ ] The software delivered for this grant will be released under an open-source license specified in the application.
- [ ] The initial PR contains only one commit (squash and force-push if needed).
- [ ] The grant will only be announced once the first milestone [has been accepted](https://github.com/w3f/Grant-Milestone-Delivery#process).

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/w3f/Grants-Program?pullRequest=870) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/w3f/Grants-Program?pullRequest=870) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/w3f/Grants-Program?pullRequest=870) it.</sub>

nodi-andy commented:
Thanks for review correction @semuelle . I made some changes in proposal and here are my responses to your review points

1. I have added a total budget (lower than initial budget) and splittet it approximately to the tasks
2. I prepared a delivery table with verification/validation conditions and documentation requirements. I will complete this table if you agree with the table. "Deliverables" were already defined in RFPS , so for my understading, the "deliverables" shall be splittet into test spec and approval conditions.
3. That is true. I have extend my sentence in chapter "development status". I will provide some gui proposals next week. This step would have been first step.

semuelle commented:
Thanks for the update, @nodi-andy. A table with verification/validation conditions and documentation requirements is great. We usually make do with the "specification" column in the deliverables tables, but this is great to have. Looking forward to your GUI proposals.

One last thing: I think the "Hide All" button referenced in the RFP refers to a single button that hides all addresses at once, not one button for each. Hence the "global" in the description.

semuelle commented:
Thanks for the update, @nodi-andy. I will share your application with the rest of the team.

nodi-andy commented:
Hi @semuelle, you are right, that the term "visibility" was over used. I will use the term "providing the address in network". Anyhow, I reworked the concept and made clear that we might implement three different visibility options for privacy groups: "address in gui", "address in network" and "account in gui".

Noc2 reviewed (changes requested):
Thanks for the grant application. I’m generally happy to support this. But could you please add at least one milestone according to our template to the grant application and the mandatory deliverables: https://github.com/w3f/Grants-Program/blob/master/applications/application-template.md#milestone-1-example--implement-substrate-modules as well as the overview section. This way we make sure that we don’t forget anything, like for example the licence of the delivery and the delivery date of the milestone. We don’t have the resources to check each deliverable individually.

takahser commented on applications/privacy-enhancement-polkadot-extension.md:
```diff
@@ -0,0 +1,119 @@
+
+### Project Details
+
+As a newcomer to Polkadot environment, i was able to create chrome extension and make some changes to work in the development environment. I was able to extend the global settings but i had to change the polkadot/ui-settings package. I didn't want to change so far but wanted to see first result so i used an existing button to see, if i can show/hide the address of an account. It was possible and  ...
```
```suggestion
As a newcomer to Polkadot environment, I was able to create a chrome extension and make some changes to work in the development environment. I was able to extend the global settings but I had to change the polkadot/ui-settings package. I didn't want to change too much but wanted to see a first result so I used an existing button to see if I could show/hide the address of an account. It was possible and I made a gif of it: https://imgur.com/a/qsfhD0z
```

takahser commented on applications/privacy-enhancement-polkadot-extension.md:
```diff
@@ -0,0 +1,119 @@
+
+As a newcomer to Polkadot environment, i was able to create chrome extension and make some changes to work in the development environment. I was able to extend the global settings but i had to change the polkadot/ui-settings package. I didn't want to change so far but wanted to see first result so i used an existing button to see, if i can show/hide the address of an account. It was possible and  ...
+
+I would make GUI mock-ups and a concept of change in the code hierarchy (interfaces, etc) before starting with the tasks of Milestone 1. In my professional projects, i like to use v-modell but i would like to present my solution and proposal to stakeholder before i start coding.
```
```suggestion
I would make GUI mock-ups and a concept of change in the code hierarchy (interfaces, etc) before starting with the tasks of Milestone 1. In my professional projects, I like to use the [v-model](https://www.geeksforgeeks.org/software-engineering-sdlc-v-model/#:~:text=The%20V%2Dmodel%20is%20a,associated%20with%20the%20testing%20phase.) but I would like to present my solution and proposal to stakeholders before I start coding.
```

takahser commented on applications/privacy-enhancement-polkadot-extension.md:
```diff
@@ -0,0 +1,119 @@
+
+### Ecosystem Fit
+
+As described in RFPS, the development will enhance the privacy of the Polkadot Extension.
```
```suggestion
As described in the [RFP](https://github.com/w3f/Grants-Program/blob/master/rfps/privacy-enhancement-polkadot-extension.md), the development will enhance the privacy of the Polkadot Extension.
```

takahser reviewed (changes requested):
@nodi-andy thanks for your application.
As pointed out by @semuelle, the application document should adhere to our [application template](https://github.com/w3f/Grants-Program/blob/master/applications/application-template.md). Currently, the following information is missing or not structured according to the template:

1. [Overview and related information](https://github.com/w3f/Grants-Program/blob/master/applications/application-template.md#overview-1)
2. The [milestone title and related information (Estimated duration, FTE, Costs)](https://github.com/w3f/Grants-Program/blob/master/applications/application-template.md#milestone-1-example--implement-substrate-modules)

Could you add these to your application?

nodi-andy commented:
Hello everyone, I was able to convert my old milestone tables into single milestones as in w3f template. I have fulfilled the first 4 Milestones and I will fulfill the others after the first review. Can you please review the first 4 milestones?

nodi-andy commented:
I had time to complete other milestones and added some more information. Independent from the milestones, I would like to have a feedback about the privacy level concept and the new three account properties: address visibility in GUI, account visibility in GUI and address/account visibility in network. This is the most crucial part. Other deliverables can be adjusted easily. Additionally, I haven't checked yet, how the test suit works and how the unit/integration tests can be implemented. I hope, that the basic testing concept is there, so I don't have to build up from scratch.

semuelle commented:
Thanks for the update, @nodi-andy. The properties list is quite helpful.

- I am not sure the distinction between "account" and "address" was intended in the RFP. If not, we could merge all the "account visibility" and "address visibility" properties into one. It might also be a bit confusing for some users to add buttons for both. 
- The RFP is unclear on whether an account can be part of more than one group. If so, a dropdown list to select groups might be insufficient. Perhaps @jonasW3F and the other reviewers can share their opinion on this. Multiple groups might get confusing, in my opinion.
- Regarding Milestone 4: the RFP suggests that hidden accounts will be listed in a separate tab. Is that your intention? It only mentions a "privacy menu".
- How are privacy levels defined? I think hiding the functionality behind an integer number might be too abstract for many users.

nodi-andy commented:
hi, here are my review responses:

- I was in the begin confused about visibility on network and gui. Now I have simplified that only accounts are managed and the accounts can be show/hide in GUI and in network (provide/conceal). The terminology can be set later.

- Assigning an account to multiple groups is not part of this concept. The gain via this feature is not worth comparing the complexity needed to realize to feature. If @jonasW3F  has another opinion, we can discuss it. I tried some taglines as experiment but didn't liked it.

- My intention is not a separate tab, but like in windows OS hiding the files. A specific account can be set as "hidded" and it is only visibile, if the option is set to "show hidden accounts". That is a handy feature, if someone does not want to user groups but wants to quick-hide any account.

- I am for saving integer values in background (and json) but I have now changed in my proposal, that the user has the options : low, medium, high for privacy level. That means, if you set the privacy for an account GUI visibility on medium and account network visibility on high, this account will be hidden in GUI, if the privacy level is medium or high. And this account will be always visible in the network, because privacy level low (int 0) can not be hidden.

I think, that most crucial question is, if we want to separated "privacy level" in to "GUI privacy level" and "Relay/Network privacy level". Other changes are minor changes and i can just adjust them as wanted.

semuelle commented:
> if you set the privacy for an account GUI visibility on medium and account network visibility on high, this account will be hidden in GUI, if the privacy level is medium or high. And this account will be always visible in the network, because privacy level low (int 0) can not be hidden.

This bit is a bit confusing. Can you elaborate once more what `low`, `medium` and `high` stand for? I would have assumed that both GUI visibility and network visibility (for lack of a better word) would be boolean values. 

Also, I just noticed that there seems to be a feature missing from the proposal that is mentioned in the RFP: `upon first authorization of a website, the extension could ask which view-groups to add it to or offer to create a new one`. This is in the specification for Milestone 5, and frankly I'd consider this quite important. In case it's not clear: the extension currently only has one button, to either authorize or not authorize the website to read the accounts from the extension. Other extensions like Metamask and Talisman will show you a checklist of accounts to select on that authorization popup alongside the button, allowing you to limit authorization to specific accounts (and in our case, ideally, also groups).

jonasW3F commented:
There is some discussion about how to approach this RFP here: https://github.com/polkadot-js/extension/issues/1037

nodi-andy commented:
@semuelle : Network and GUI visibilities are boolean, which are decided based on the privacy level of account and current privacy level. Basically: "thisAccount.networkVisibility = this.Account.privacyLevel >= Settings.currentPrivacyLevel" but with the new discussion https://github.com/polkadot-js/extension/issues/1037 the RFP goals and Milestones incl. Milestone 5 might be changed, see below.

@jonasW3F : Do you think that RFP shall be changed as proposed here:  https://github.com/polkadot-js/extension/issues/1037#issuecomment-1081951024

Assigning the accounts to the groups shall be removed from RFP? Also means, Milestone 2 shall be removed?

I think, what @Tbaut tries to explain with "lists", is the same as assigning an account to a group. Adding accounts to list and assigning an account to a group is technically the same operation.

To clarify my proposal about "privacy levels: low, medium, high". I would like to extend this idea that a dapp website shall be also linked to the privacy level. Default; if the first autherization call is made, the web page is classified as "privacy level: low", so the webpage can only see the accounts which are classfied as "privacy level: low". If the user trusts the webpage or want to share more account, so the webpage can be classified as "privacy level: medium". We can remove the grouping of accounts, if the grouping is unwanted, and each account has a privacy level. If a user is not interessed setting the privacy level, the user will not even note that such a features exists.

The proposal of @jacogr, that on top of privacy shall be possible to show/hide account for dapp websites was part of my concept.

I will need some time to change the concept and might response next week.

nodi-andy commented:
@semuelle : Network and GUI visibilities are boolean, which are decided based on the privacy level of account and current privacy level. Basically: "thisAccount.networkVisibility = thisAccount.privacyLevel >= Settings.currentPrivacyLevel" but with the new discussion https://github.com/polkadot-js/extension/issues/1037 the RFP goals and Milestones incl. Milestone 5 might be changed, see below.

@jonasW3F : Do you think that RFP shall be changed as proposed here:  https://github.com/polkadot-js/extension/issues/1037#issuecomment-1081951024

Assigning the accounts to the groups shall be removed from RFP? Also means, Milestone 2 shall be removed?

I think, what @Tbaut tries to explain with "lists", is the same as assigning an account to a group. Adding accounts to list and assigning an account to a group is technically the same operation.

To clarify my proposal about "privacy levels: low, medium, high". I would like to extend this idea that a dapp website shall be also linked to the privacy level. Default; if the first autherization call is made, the web page is classified as "privacy level: low", so the webpage can only see the accounts which are classfied as "privacy level: low". If the user trusts the webpage or want to share more account, so the webpage can be classified as "privacy level: medium". We can remove the grouping of accounts, if the grouping is unwanted, and each account has a privacy level. If a user is not interessed setting the privacy level, the user will not even note that such a features exists.

The proposal of @jacogr, that on top of privacy shall be possible to show/hide account for dapp websites was part of my concept.

I will need some time to change the concept and might response next week.

nodi-andy commented:
@semuelle : Network and GUI visibilities are boolean, which are decided based on the privacy level of account and current privacy level. Basically: "thisAccount.networkVisibility = thisAccount.privacyLevel >= Settings.currentPrivacyLevel" but with the new discussion https://github.com/polkadot-js/extension/issues/1037 the RFP goals and Milestones incl. Milestone 5 might be changed, see below.

@jonasW3F : Do you think that RFP shall be changed as proposed here:  https://github.com/polkadot-js/extension/issues/1037#issuecomment-1081951024

Assigning the accounts to the groups shall be removed from RFP? Also means, Milestone 2 shall be removed?

I think, what @Tbaut tries to explain with "lists", is the same as assigning an account to a group. Adding accounts to list and assigning an account to a group is technically the same operation.

To clarify my proposal about "privacy levels: low, medium, high". I would like to extend this idea that a dapp website shall be also linked to the privacy level. Default; if the first autherization call is made, the web page is classified as "privacy level: low", so the webpage can only see the accounts which are classfied as "privacy level: low". If the user trusts the webpage or want to share more account, so the webpage can be classified as "privacy level: medium". We can remove the grouping of accounts, if the grouping is unwanted, and each account has a privacy level. If a user is not interessed setting the privacy level, the user will not even note that such a features exists.

The proposal of @jacogr, that on top of privacy shall be possible to show/hide account for dapp websites was part of my concept.

I will need some time to change the concept and might response next week.

nodi-andy commented:
I have changed the concept and documented the concept in the first step without the specification and milestone to agree on the goals. Can someone please review the new chapters "State of Art", "Privacy Concept" and "Implementation". This concept do not cover "account groups" and "privacy level" but shall fulfill the user expectations as discussed here: https://github.com/polkadot-js/extension/issues/1037

I would like to make more visuals, better MD formatting and UML diagramms but i am middle of Amazonas in South America and have very restricted internet, so I hope for more collaboration in 2 weeks after i am back to Europe.

Noc2 reviewed (changes requested):
For reference: https://github.com/polkadot-js/extension/issues/1037#issuecomment-1084283229

nodi-andy commented:
> For reference: [polkadot-js/extension#1037 (comment)](https://github.com/polkadot-js/extension/issues/1037#issuecomment-1084283229)

Thanks for your feedback @Noc2

Noc2 reviewed (changes requested):
@nodi-andy  Thanks again for all the effort you put into this. However we decided to go ahead with https://github.com/w3f/Grants-Program/pull/908 instead of this application. Therefore, I’m going to close this PR. 

That said, we would be more than happy if you work on something else in our ecosystem and we are generally interested in more UI/UX related projects. I personally would also be happy to jump on a call with you and brainstorm other ideas. Feel free to send me an email to ​​david@web3.foundation.
